pallets/xyk/src/lib.rs

```diff
@@ -11,92 +8,285 @@ mod mock;
 
     // Errors inform users that something went wrong.
     #[pallet::error]
     pub enum Error<T> {
-        /// Error names should be descriptive.
-        NoneValue,
-        /// Errors should have helpful documentation associated with them.
-        StorageOverflow,
+        /// No liquidity available
+        NoLiquidity,
+        /// Not enough tokens
+        NotEnoughTokens,
+        LiquidityAlreadyExists,
+    }
+
+    impl<T: Config> Pallet<T> {
+        fn transfer(
+            from: T::AccountId,
+            to: T::AccountId,
+            asset_id: AssetId<T>,
+            amount: Balance<T>,
+        ) -> Result<(), Error<T>> {
+            pallet_assets::Pallet::<T>::transfer(
+                RawOrigin::Signed(from).into(),
+                asset_id.into(),
+                T::Lookup::unlookup(to),
+                amount,
+            )
+            .map_err(|_| Error::<T>::NotEnoughTokens)
+        }
+
+        fn mint_lp(lp: AssetId<T>, amount: Balance<T>, to: T::AccountId) {
+            pallet_assets::Pallet::<T>::mint(
+                RawOrigin::Signed(T::DEXAddr::get()).into(),
+                lp.into(),
+                T::Lookup::unlookup(to),
+                amount,
+            )
+            .unwrap()
+        }
+
+        fn burn_lp(lp: AssetId<T>, amount: Balance<T>, from: T::AccountId) {
+            pallet_assets::Pallet::<T>::burn(
+                RawOrigin::Signed(T::DEXAddr::get()).into(),
+                lp.into(),
+                T::Lookup::unlookup(from),
+                amount,
+            )
+            .unwrap()
+        }
+
+        pub fn lp_token(x: AssetId<T>, y: AssetId<T>) -> AssetId<T> {
+            let (x, y) = if x > y { (y, x) } else { (x, y) };
+            AssetId::<T>::decode(&mut T::AssetIdHash::hash_of(&(x, y)).encode().as_ref()).unwrap()
+        }
+
+        pub fn _swap(
+            origin: T::AccountId,
+            (from_id, from): (AssetId<T>, Balance<T>),
+            to_id: AssetId<T>,
+        ) -> Result<Balance<T>, Error<T>> {
+            let lp = Self::lp_token(from_id, to_id);
+
+            LPPairs::<T>::mutate(lp, |p| {
+                let p = p.as_mut().ok_or(Error::<T>::NoLiquidity)?;
+
+                Self::transfer(origin.clone(), T::DEXAddr::get(), from_id, from)?;
+
+                // x * y = k
+                // (x + dx) * (y - dy) = k
+                // y - dy = x * y / (x + dx)
+                // dy = y - x * y / (x + dx)
+                let new_x = p.x.1 + from;
+                let new_y = (p.x.1 * p.y.1) / new_x;
+
+                let dy = p.y.1 - new_y;
+
+                Self::transfer(T::DEXAddr::get(), origin.clone(), to_id, dy)
+                    .expect("Should always work by because of x*y=k architecture");
+
+                p.x.1 = new_x;
+                p.y.1 = new_y;
+
+                Ok(dy)
+            })
+        }
+
+        fn _create_lp(
+            origin: T::AccountId,
+            (x_id, x): (AssetId<T>, Balance<T>),
```

Comment:
Again, same concept:
```suggestion
            (x_id, x_balance): (AssetId<T>, Balance<T>),
```